### PR TITLE
Interpolation search levels

### PR DESCRIPTION
**Pull request by NaderMahdi**
* **Added interpolation search levels for better speed/quality trade-offs**

    - enc_mode = ENC_M0
       interpolation_search_level = IT_SEARCH_FULL_LOOP
    - Otherwise
       interpolation_search_level = IT_SEARCH_OFF

    // Interpolation search Level       Settings
    // 0                                              OFF
    // 1                                              Interpolation search at inter-depth
    // 2                                              Interpolation search at full loop
    // 3                                              Interpolation search at fast loop

### CODE REVIEW

File: Source/Lib/Codec/EbProductCodingLoop.c [hunk 1]
```diff
@@ -2001,17 +2001,35 @@ static void CflPrediction(
     }
 }
 
+#if TX_SEARCH_LEVELS
+uint8_t get_skip_tx_search_flag(
+    int32_t                  sq_size,
+    uint64_t                 ref_fast_cost,
+    uint64_t                 cu_cost,
+    uint64_t                 weight
```
Comment (hassount) [inline Source/Lib/Codec/EbProductCodingLoop.c]: coding style:
uint64_t                 weight){

File: Source/Lib/Codec/EbProductCodingLoop.c [hunk 2]
```diff
@@ -2441,6 +2471,312 @@ EbBool allowed_ns_cu(
...
+
+        if (candidate_ptr->type == INTRA_MODE)
+            context_ptr->md_local_cu_unit[cu_ptr->mds_idx].cost_luma = candidateBuffer->full_cost_luma;
+
+
+        context_ptr->md_ep_pipe_sb[cu_ptr->mds_idx].merge_cost = *candidateBuffer->full_cost_merge_ptr;
+        context_ptr->md_ep_pipe_sb[cu_ptr->mds_idx].skip_cost = *candidateBuffer->full_cost_skip_ptr;
+
+
+        if (candidate_ptr->type == INTER_MODE && candidate_ptr->merge_flag == EB_TRUE) {
```
Comment (hassount) [inline Source/Lib/Codec/EbProductCodingLoop.c]: single line if statements should not have brackets

File: Source/Lib/Codec/EbProductCodingLoop.c [hunk 3]
```diff
@@ -2441,6 +2471,312 @@ EbBool allowed_ns_cu(
...
+                uint32_t  bheight = context_ptr->blk_geom->tx_height[txb_itr] < 64 ? context_ptr->blk_geom->tx_height[txb_itr] : 32;
+
+                int32_t* srcPtr = &(((int32_t*)buffer_ptr_array[lowestCostIndex]->residualQuantCoeffPtr->bufferY)[txb_1d_offset]);
+                int32_t* dstPtr = &(((int32_t*)context_ptr->cu_ptr->coeff_tmp->bufferY)[txb_1d_offset]);
+
+                uint32_t j;
+
+                for (j = 0; j < bheight; j++)
+                {
+                    memcpy(dstPtr + j * bwidth, srcPtr + j * bwidth, bwidth * sizeof(int32_t));
```
Comment (hassount) [inline Source/Lib/Codec/EbProductCodingLoop.c]: we should use EB_MEMCPY

Comment (NaderMahdi) [reply Source/Lib/Codec/EbProductCodingLoop.c]: This is a part of debugging code that will eventually be removed.

File: Source/Lib/Codec/EbProductCodingLoop.c (hunk 2 above)
Comment (NaderMahdi) [reply Source/Lib/Codec/EbProductCodingLoop.c]: Done

File: Source/Lib/Codec/EbProductCodingLoop.c [hunk 4]
```diff
@@ -2441,6 +2481,313 @@ EbBool allowed_ns_cu(
...
+        memset(candidate_ptr->eob[2], 0, sizeof(uint16_t)*MAX_TXB_COUNT);
+        memset(count_non_zero_coeffs[2], 0, sizeof(uint32_t)*MAX_NUM_OF_TU_PER_CU);
+
+
+        ProductFullLoopTxSearch(
+            candidateBuffer,
+            context_ptr,
+            picture_control_set_ptr
+        );
+
```
Comment (andrey-norkin) [inline Source/Lib/Codec/EbProductCodingLoop.c]: Nit: sparse code

File: Source/Lib/Codec/EbProductCodingLoop.c [hunk 5]
```diff
@@ -2441,6 +2481,313 @@ EbBool allowed_ns_cu(
...
+                        memcpy(dstPtr + j * bwidth, srcPtr + j * bwidth, bwidth * sizeof(int32_t));
+                    }
+                }
+
+                txb_1d_offset += context_ptr->blk_geom->tx_width[txb_itr] * context_ptr->blk_geom->tx_height[txb_itr];
+                if (context_ptr->blk_geom->has_uv)
+                    txb_1d_offset_uv += context_ptr->blk_geom->tx_width_uv[txb_itr] * context_ptr->blk_geom->tx_height_uv[txb_itr];
+
+            }
+
```
Comment (andrey-norkin) [inline Source/Lib/Codec/EbProductCodingLoop.c]: Nit: sparse code

File: Source/Lib/Codec/EbProductCodingLoop.c (hunk 4 above)
Comment (NaderMahdi) [reply Source/Lib/Codec/EbProductCodingLoop.c]: Done.

File: Source/Lib/Codec/EbProductCodingLoop.c (hunk 5 above)
Comment (NaderMahdi) [reply Source/Lib/Codec/EbProductCodingLoop.c]: This code is used for debugging under NO_ENCDEC. It will be removed when the NO_ENCDEC cleaned up.

Comment (NaderMahdi) [reply Source/Lib/Codec/EbProductCodingLoop.c]: code under NO_ENCDEC for debugging which eventually will be removed as part of the next cleanup part.

**Comment by NaderMahdi** [comment]
@andrey-norkin - Any objection to merge this PR?

**Pull request merged**